lightning/src/routing/router.rs

```diff
@@ -4991,17 +4992,33 @@ mod benches {
 				let payee = Payee::from_node_id(dst).with_features(features.clone());
 				let first_hop = first_hop(src);
 				let amt = seed as u64 % 1_000_000;
-				if get_route(&payer, &payee, &graph, Some(&[&first_hop]), amt, 42, &DummyLogger{}, &scorer).is_ok() {
-					path_endpoints.push((first_hop, payee, amt));
+				if let Ok(route) = get_route(&payer, &payee, &graph, Some(&[&first_hop]), amt, 42, &DummyLogger{}, &scorer) {
+					routes.push(route);
+					route_endpoints.push((first_hop, payee, amt));
 					continue 'load_endpoints;
 				}
 			}
 		}
 
+		// ...and seed the scorer with success and failure data...
+		for route in routes {
+			let amount = route.get_total_amount();
+			if amount < 250_000 {
+				for path in route.paths {
+					scorer.payment_path_successful(&path.iter().collect::<Vec<_>>());
+				}
+			} else if amount > 750_000 {
+				for path in route.paths {
+					let short_channel_id = path[amount as usize % path.len()].short_channel_id;
+					scorer.payment_path_failed(&path.iter().collect::<Vec<_>>(), short_channel_id);
+				}
+			}
```

Comment:
@TheBlueMatt So, when the `ProbabilisticScorer` is seeded in this manner, `get_route` can't find a route between the payer and payee below when the amount is above `750_000`, it seems (i.e. when a channel would get a `u64::max_value` penalty). It does work when the failed `short_channel_id` is set here as `path[path.len() / 2]`, though. So maybe something in the routing is not working if a maximum penalty is near the payee? Will continue to debug, but let me know if something pops out.

Reply:
Isn't this teaching the scorer that there is not 750k sats available to route? Presumably if the channel that we teach the scorer has < 750k sats is in the middle the router will either find another path or build an MPP but if its near the end there may only be one or two channels available?

Reply:
I might be mistaken, but I thought available to route is different from cost to route, the latter being fees plus penalty. That is, even if every channel had a cost of `u64::max_value`, we would ultimately find some route if there were enough availability.

Reply:
So, interesting data point: When I switched `ZeroPenaltyScorer` to always return `u64:max_value`, the benchmark is pinned at 100% and hasn't finished in over 10 minutes.

Reply:
> I might be mistaken, but I thought available to route is different from cost to route, the latter being fees plus penalty. That is, even if every channel had a cost of u64::max_value, we would ultimately find some route if there were enough availability.

If a cost is `u64::max_value` the channel will be ignored - the fee for the path is considered to have overflowed and we cannot possibly pay an infinite fee.

> So, interesting data point: When I switched ZeroPenaltyScorer to always return u64:max_value, the benchmark is pinned at 100% CPU and hasn't finished in over 10 minutes.

I think this makes sense, I assume its stuck in the setup loop looking for routes it can actually take to feed the benchmarker.

Reply:
> If a cost is `u64::max_value` the channel will be ignored - the fee for the path is considered to have overflowed and we cannot possibly pay an infinite fee.
>

Hmm... could you point me to where the overflow for the fee occurs? I thought the penalty was considered only in terms of a cost, not ability to pay. Looking bit closer, it may be because the infinite cost is not less than the also-infinite-first-initialized old cost:

https://github.com/lightningdevkit/rust-lightning/blob/7b6a7bbc722356e1ae6b196b95755424ea82b000/lightning/src/routing/router.rs#L930

https://github.com/lightningdevkit/rust-lightning/blob/7b6a7bbc722356e1ae6b196b95755424ea82b000/lightning/src/routing/router.rs#L843
 
> I think this makes sense, I assume its stuck in the setup loop looking for routes it can actually take to feed the benchmarker.

Ah, yeah, that makes sense.

Reply:
> Looking bit closer, it may be because the infinite cost is not less than the also-infinite-first-initialized old cost:

Yep, that sounds right, it'll overflow when we add it to the `$next_hops_path_penalty_msat`, setting `path_penalty_msat` to `u64::max_value()` which will never be less than the `old_cost` - this is correct, returning `u64::max_value()` for a penalty *should* disable the channel, imo.

Reply:
Changed to select the middle hop as the failed one.